core/desugarFuns.ml

```diff
@@ -92,22 +100,26 @@ object (o : 'self_type)
   method private desugarFunLit argss lin lam location =
     let inner_mb     = snd (last argss) in
     let (o, lam, rt) = o#funlit inner_mb lam in
-    let ft = List.fold_right (fun (args, mb) rt ->
+    let fti = List.fold_right (fun (args, mb) rt ->
                  if DeclaredLinearity.is_linear lin
                  then `Lolli (args, mb, rt)
                  else `Function (args, mb, rt))
                argss rt in
     let f = gensym ~prefix:"_fun_" () in
-    let body, tyvars, ft, fti =
-      (* FIXME: do the generalisation and instantiation steps as part
-         of type inference and store the results in the FunLit for use
-         here. (The current implementation is broken because
-         generalisation turns previously flexible type variables into
-         rigid ones, which persist from the first round of type
-         checking.) *)
-      let (tyvars, _tyargs), ft = Generalise.generalise env.Types.var_env ft in
-      let (tyargs, fti) = Instantiate.typ ft in
-      tappl (Var f, tyargs), tyvars, ft, fti in
+    (* FIXME: do the generalisation and instantiation steps as part of type
```

Comment:
So on thing @slindley and I realised is that we don't actually need to generalise here at all - it's perfectly fine just to have a named function with no type arguments.

However, we do need a way to prevent the type checker from generalising this function - we're probably going to introduce a "frozen let" (well, frozen function binding), which basically enforces that the quantifiers don't change.

Reply:
OK. Shall we hold off on merging this until you've implemented that, in that case?

Reply:
Yep, I think so. I've got the bare bones working, and it seems to solve the problem correctly, just need to polish it off :).